When target certificate is not accepted, requests fail with 404

When attempting to connect to a target application with JMX+SSL enabled but the certificate is not trusted by ContainerJFR, requests fail with a 404 response status code. This perhaps reflects the "non-JRMP server found at address" (paraphrased) exception message seen by ContainerJFR itself, but it is not an appropriate status code for ContainerJFR to include in a response for this type of failure.

I think ideally the response code, or perhaps some detail included in the response body, should be identifiable for the `web-client` to detect and then display a useful error message or prompt to the user, informing them of the cert trust issue and perhaps guiding them to add the cert to ContainerJFR's truststore.

```
[WARN] io.vertx.ext.web.handler.impl.HttpStatusException: Not Found
Caused by: org.openjdk.jmc.rjmx.ConnectionException caused by java.io.IOException: Failed to retrieve RMIServer stub: javax.naming.CommunicationException [Root exception is java.rmi.ConnectIOException: non-JRMP server at remote endpoint]
	at org.openjdk.jmc.rjmx.internal.RJMXConnection.connect(RJMXConnection.java:441)
	at com.redhat.rhjmc.containerjfr.core.net.JFRConnection.attemptConnect(JFRConnection.java:212)
	at com.redhat.rhjmc.containerjfr.core.net.JFRConnection.connect(JFRConnection.java:169)
	at com.redhat.rhjmc.containerjfr.core.net.JFRConnection.getService(JFRConnection.java:109)
	at com.redhat.rhjmc.containerjfr.net.web.http.api.v1.TargetEventsGetHandler.lambda$handleAuthenticated$0(TargetEventsGetHandler.java:97)
	at com.redhat.rhjmc.containerjfr.net.TargetConnectionManager.executeConnectedTask(TargetConnectionManager.java:87)
	at com.redhat.rhjmc.containerjfr.net.web.http.api.v1.TargetEventsGetHandler.handleAuthenticated(TargetEventsGetHandler.java:93)
	at com.redhat.rhjmc.containerjfr.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:82)
	at com.redhat.rhjmc.containerjfr.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:61)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1034)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:131)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:133)
	at io.vertx.ext.web.handler.impl.StaticHandlerImpl.sendStatic(StaticHandlerImpl.java:175)
	at io.vertx.ext.web.handler.impl.StaticHandlerImpl.handle(StaticHandlerImpl.java:145)
	at io.vertx.ext.web.handler.impl.StaticHandlerImpl.handle(StaticHandlerImpl.java:51)
	at com.redhat.rhjmc.containerjfr.net.web.http.generic.StaticAssetsGetHandler.handle(StaticAssetsGetHandler.java:79)
	at com.redhat.rhjmc.containerjfr.net.web.http.generic.StaticAssetsGetHandler.handle(StaticAssetsGetHandler.java:53)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1034)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:131)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:133)
	at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:54)
	at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:36)
	at com.redhat.rhjmc.containerjfr.net.web.WebServer.lambda$start$4(WebServer.java:203)
	at com.redhat.rhjmc.containerjfr.net.HttpServer$HandlerDelegate.handle(HttpServer.java:166)
	at io.vertx.core.http.impl.WebSocketRequestHandler.handle(WebSocketRequestHandler.java:50)
	at io.vertx.core.http.impl.WebSocketRequestHandler.handle(WebSocketRequestHandler.java:32)
	at io.vertx.core.http.impl.Http1xServerConnection.handleMessage(Http1xServerConnection.java:136)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:366)
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:43)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:229)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:173)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.channelRead(WebSocketServerExtensionHandler.java:101)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.vertx.core.http.impl.Http1xUpgradeToH2CHandler.channelRead(Http1xUpgradeToH2CHandler.java:109)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:271)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1518)
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1267)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1314)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:501)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:440)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.io.IOException: Failed to retrieve RMIServer stub: javax.naming.CommunicationException [Root exception is java.rmi.ConnectIOException: non-JRMP server at remote endpoint]
	at java.management.rmi/javax.management.remote.rmi.RMIConnector.connect(RMIConnector.java:370)
	at org.openjdk.jmc.rjmx.internal.RJMXConnection.connectJmxConnector(RJMXConnection.java:487)
	at org.openjdk.jmc.rjmx.internal.RJMXConnection.establishConnection(RJMXConnection.java:464)
	at org.openjdk.jmc.rjmx.internal.RJMXConnection.connect(RJMXConnection.java:434)
	... 78 more
Caused by: javax.naming.CommunicationException [Root exception is java.rmi.ConnectIOException: non-JRMP server at remote endpoint]
	at jdk.naming.rmi/com.sun.jndi.rmi.registry.RegistryContext.lookup(RegistryContext.java:137)
	at java.naming/com.sun.jndi.toolkit.url.GenericURLContext.lookup(GenericURLContext.java:207)
	at java.naming/javax.naming.InitialContext.lookup(InitialContext.java:409)
	at java.management.rmi/javax.management.remote.rmi.RMIConnector.findRMIServerJNDI(RMIConnector.java:1839)
	at java.management.rmi/javax.management.remote.rmi.RMIConnector.findRMIServer(RMIConnector.java:1813)
	at java.management.rmi/javax.management.remote.rmi.RMIConnector.connect(RMIConnector.java:302)
	... 81 more
Caused by: java.rmi.ConnectIOException: non-JRMP server at remote endpoint
	at java.rmi/sun.rmi.transport.tcp.TCPChannel.createConnection(TCPChannel.java:244)
	at java.rmi/sun.rmi.transport.tcp.TCPChannel.newConnection(TCPChannel.java:196)
	at java.rmi/sun.rmi.server.UnicastRef.newCall(UnicastRef.java:343)
	at java.rmi/sun.rmi.registry.RegistryImpl_Stub.lookup(RegistryImpl_Stub.java:116)
	at jdk.naming.rmi/com.sun.jndi.rmi.registry.RegistryContext.lookup(RegistryContext.java:133)
	... 86 more

[INFO] (127.0.0.1:58360): GET /api/v1/targets/service%3Ajmx%3Armi%3A%2F%2F%2Fjndi%2Frmi%3A%2F%2Fcontainer-jfr%3A9093%2Fjmxrmi/events 404 70ms
```

--- a/HTTP_API.md
+++ b/HTTP_API.md
@@ -475,10 +475,6 @@
     `501` - The Grafana datasource URL is malformed.
     The body is an error message.
 
-    `502` - Container JFR received an invalid response from the
-    Grafana datasource server after sending the upload request.
-    The body is an error message.
-
     ###### example
     ```
     $ curl -X POST localhost:8181/api/v1/recordings/localhost_foo_20200911T144545Z.jfr/upload
@@ -553,6 +549,10 @@
 
     `500` - There was an unexpected error. The body is an error message.
 
+    `502` - JMX connection failed. This is generally because the target
+    application has SSL enabled over JMX, but ContainerJFR does not trust the
+    certificate.
+
     ###### example
     ```
     $ curl localhost:8181/api/v1/targets/localhost/events --output events
@@ -595,6 +595,10 @@
 
     `500` - There was an unexpected error. The body is an error message.
 
+    `502` - JMX connection failed. This is generally because the target
+    application has SSL enabled over JMX, but ContainerJFR does not trust the
+    certificate.
+
     ###### example
     ```
     $ curl -X DELETE localhost:8181/api/v1/targets/localhost/recordings/foo
@@ -633,6 +637,10 @@
     `500` - The recording was found but it could not be written to the
     response. Or there was an unexpected error. The body is an error message.
 
+    `502` - JMX connection failed. This is generally because the target
+    application has SSL enabled over JMX, but ContainerJFR does not trust the
+    certificate.
+
     ###### example
     ```
     $ curl localhost:8181/api/v1/targets/localhost/recordings/foo --output foo.jfr
@@ -667,6 +675,10 @@
     the authentication scheme that is used.
 
     `500` - There was an unexpected error. The body is an error message.
+
+    `502` - JMX connection failed. This is generally because the target
+    application has SSL enabled over JMX, but ContainerJFR does not trust the
+    certificate.
 
     ###### example
     ```
@@ -718,6 +730,10 @@
     the authentication scheme that is used.
 
     `500` - There was an unexpected error. The body is an error message.
+
+    `502` - JMX connection failed. This is generally because the target
+    application has SSL enabled over JMX, but ContainerJFR does not trust the
+    certificate.
 
     ###### example
     ```
@@ -771,6 +787,10 @@
 
     `500` - There was an unexpected error. The body is an error message.
 
+    `502` - JMX connection failed. This is generally because the target
+    application has SSL enabled over JMX, but ContainerJFR does not trust the
+    certificate.
+
     **STOP**
 
     `200` - No body.
@@ -821,6 +841,10 @@
     the authentication scheme that is used.
 
     `500` - There was an unexpected error. The body is an error message.
+
+    `502` - JMX connection failed. This is generally because the target
+    application has SSL enabled over JMX, but ContainerJFR does not trust the
+    certificate.
 
     ###### example
     ```
@@ -889,6 +913,10 @@
 
     `500` - There was an unexpected error. The body is an error message.
 
+    `502` - JMX connection failed. This is generally because the target
+    application has SSL enabled over JMX, but ContainerJFR does not trust the
+    certificate.
+
     ###### example
     ```
     $ curl --data "recordingName=foo&duration=5&events=template=ALL" localhost:8181/api/v1/targets/localhost/recordings
@@ -937,6 +965,10 @@
     Grafana datasource after sending the upload request.
     The body is an error message.
 
+    `502` - JMX connection failed. This is generally because the target
+    application has SSL enabled over JMX, but ContainerJFR does not trust the
+    certificate.
+
     ###### example
     ```
     $ curl -X POST localhost:8181/api/v1/targets/localhost/recordings/foo/upload
@@ -976,6 +1008,10 @@
 
     `500` - There was an unexpected error. The body is an error message.
 
+    `502` - JMX connection failed. This is generally because the target
+    application has SSL enabled over JMX, but ContainerJFR does not trust the
+    certificate.
+
     ###### example
     ```
     $ curl localhost:8181/api/v1/targets/localhost/reports/foo --output report.html
@@ -1013,6 +1049,10 @@
     the authentication scheme that is used.
 
     `500` - There was an unexpected error. The body is an error message.
+
+    `502` - JMX connection failed. This is generally because the target
+    application has SSL enabled over JMX, but ContainerJFR does not trust the
+    certificate.
 
     ###### example
     ```
@@ -1053,6 +1093,10 @@
 
     `500` - There was an unexpected error. The body is an error message.
 
+    `502` - JMX connection failed. This is generally because the target
+    application has SSL enabled over JMX, but ContainerJFR does not trust the
+    certificate.
+
     ###### example
     ```
     $ curl localhost:8181/api/v1/targets/localhost/templates/Profiling/type/TARGET --output Continuous.jfc
@@ -1091,6 +1135,10 @@
     the authentication scheme that is used.
 
     `500` - There was an unexpected error. The body is an error message.
+
+    `502` - JMX connection failed. This is generally because the target
+    application has SSL enabled over JMX, but ContainerJFR does not trust the
+    certificate.
 
     ###### example
     ```
@@ -1211,6 +1259,10 @@
     the authentication scheme that is used.
 
     `500` - There was an unexpected error. The body is an error message.
+
+    `502` - JMX connection failed. This is generally because the target
+    application has SSL enabled over JMX, but ContainerJFR does not trust the
+    certificate.
 
     ###### example
     ```

--- a/HTTP_API.md
+++ b/HTTP_API.md
@@ -963,11 +963,9 @@
 
     `502` - Container JFR received an invalid response from the
     Grafana datasource after sending the upload request.
-    The body is an error message.
-
-    `502` - JMX connection failed. This is generally because the target
-    application has SSL enabled over JMX, but ContainerJFR does not trust the
-    certificate.
+    The body is an error message. Or, the JMX connection failed. This is
+    generally because the target application has SSL enabled over JMX, but
+    ContainerJFR does not trust the certificate.
 
     ###### example
     ```

--- a/src/main/java/com/redhat/rhjmc/containerjfr/net/web/http/AbstractAuthenticatedRequestHandler.java
+++ b/src/main/java/com/redhat/rhjmc/containerjfr/net/web/http/AbstractAuthenticatedRequestHandler.java
@@ -42,10 +42,13 @@
 package com.redhat.rhjmc.containerjfr.net.web.http;
 
 import java.nio.charset.StandardCharsets;
+import java.rmi.ConnectIOException;
 import java.util.Base64;
 import java.util.concurrent.Future;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
+
+import org.apache.commons.lang3.exception.ExceptionUtils;
 
 import org.openjdk.jmc.rjmx.ConnectionException;
 
@@ -86,9 +89,13 @@ public abstract class AbstractAuthenticatedRequestHandler implements RequestHand
             Throwable cause = e.getCause();
             if (cause instanceof SecurityException) {
                 ctx.response().putHeader(JMX_AUTHENTICATE_HEADER, "Basic");
-                throw new HttpStatusException(427, e);
+                throw new HttpStatusException(427, "JMX Authentication Failure", e);
             }
-            throw new HttpStatusException(404, e);
+            Throwable rootCause = ExceptionUtils.getRootCause(e);
+            if (rootCause instanceof ConnectIOException) {
+                throw new HttpStatusException(502, "Target SSL Untrusted", e);
+            }
+            throw new HttpStatusException(500, e);
         } catch (Exception e) {
             throw new HttpStatusException(500, e.getMessage(), e);
         }

--- a/src/test/java/com/redhat/rhjmc/containerjfr/net/web/http/AbstractAuthenticatedRequestHandlerTest.java
+++ b/src/test/java/com/redhat/rhjmc/containerjfr/net/web/http/AbstractAuthenticatedRequestHandlerTest.java
@@ -43,6 +43,7 @@ package com.redhat.rhjmc.containerjfr.net.web.http;
 
 import static org.mockito.Mockito.when;
 
+import java.rmi.ConnectIOException;
 import java.util.concurrent.CompletableFuture;
 
 import org.hamcrest.MatcherAssert;
@@ -124,13 +125,13 @@ class AbstractAuthenticatedRequestHandlerTest {
         }
 
         @Test
-        void shouldThrow404IfConnectionFails() {
+        void shouldThrow500IfConnectionFails() {
             Exception expectedException = new ConnectionException("");
             handler = new ThrowingAuthenticatedHandler(auth, expectedException);
 
             HttpStatusException ex =
                     Assertions.assertThrows(HttpStatusException.class, () -> handler.handle(ctx));
-            MatcherAssert.assertThat(ex.getStatusCode(), Matchers.equalTo(404));
+            MatcherAssert.assertThat(ex.getStatusCode(), Matchers.equalTo(500));
         }
 
         @Test
@@ -146,6 +147,19 @@ class AbstractAuthenticatedRequestHandlerTest {
                     Assertions.assertThrows(HttpStatusException.class, () -> handler.handle(ctx));
             MatcherAssert.assertThat(ex.getStatusCode(), Matchers.equalTo(427));
             Mockito.verify(resp).putHeader("X-JMX-Authenticate", "Basic");
+        }
+
+        @Test
+        void shouldThrow502IfConnectionFailsDueToSslTrust() {
+            Exception cause = new ConnectIOException("SSL trust");
+            Exception expectedException = new ConnectionException("");
+            expectedException.initCause(cause);
+            handler = new ThrowingAuthenticatedHandler(auth, expectedException);
+
+            HttpStatusException ex =
+                    Assertions.assertThrows(HttpStatusException.class, () -> handler.handle(ctx));
+            MatcherAssert.assertThat(ex.getStatusCode(), Matchers.equalTo(502));
+            MatcherAssert.assertThat(ex.getPayload(), Matchers.equalTo("Target SSL Untrusted"));
         }
 
         @Test

